terraform/jenkins/binary_cache.tf

```diff
@@ -0,0 +1,115 @@
+# SPDX-FileCopyrightText: 2023 Technology Innovation Institute (TII)
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module "binary_cache_image" {
+  source = "../../tf-modules/azurerm-nix-vm-image"
+
+  nix_attrpath   = "outputs.nixosConfigurations.binary-cache.config.system.build.azureImage"
+  nix_entrypoint = "${path.module}/../.."
+
+
+  name                = "binary-cache"
+  resource_group_name = azurerm_resource_group.default.name
+  location            = azurerm_resource_group.default.location
+
+  storage_account_name   = azurerm_storage_account.vm_images.name
+  storage_container_name = azurerm_storage_container.vm_images.name
+}
+
+module "binary_cache_vm" {
+  source = "../../tf-modules/azurerm-linux-vm"
+
+  resource_group_name = azurerm_resource_group.default.name
+  location            = azurerm_resource_group.default.location
+
+  virtual_machine_name         = "ghaf-binary-cache"
+  virtual_machine_size         = "Standard_D1_v2"
+  virtual_machine_source_image = module.binary_cache_image.image_id
+
+  virtual_machine_custom_data = join("\n", ["#cloud-config", yamlencode({
+    users = [
+      for user in toset(["bmg", "flokli", "hrosten"]) : {
+        name                = user
+        sudo                = "ALL=(ALL) NOPASSWD:ALL"
+        ssh_authorized_keys = local.ssh_keys[user]
+      }
+    ]
+    # mount /dev/disk/by-lun/10 to /var/lib/caddy
+    disk_setup = {
+      "/dev/disk/by-lun/10" = {
+        layout  = false # don't partition
+        timeout = 60    # wait for device to appear
+      }
+    }
+    fs_setup = [
+      {
+        filesystem = "ext4"
+        partition  = "auto"
+        device     = "/dev/disk/by-lun/10"
+        label      = "caddy"
+      }
+    ]
+    mounts = [
+      ["/dev/disk/by-label/caddy", "/var/lib/caddy"]
+    ]
+  })])
+
+  subnet_id = azurerm_subnet.binary_cache.id
+}
+
+resource "azurerm_subnet" "binary_cache" {
+  name                 = "ghaf-infra-binary-cache"
+  resource_group_name  = azurerm_resource_group.default.name
+  virtual_network_name = azurerm_virtual_network.vnet.name
+  address_prefixes     = ["10.0.3.0/28"]
+}
+
+# Allow inbound HTTP(S)
+resource "azurerm_network_interface_security_group_association" "binary_cache_vm" {
+  network_interface_id      = module.binary_cache_vm.virtual_machine_network_interface_id
+  network_security_group_id = azurerm_network_security_group.binary_cache_vm.id
+}
+
+resource "azurerm_network_security_group" "binary_cache_vm" {
+  name                = "binary-cache-vm"
+  resource_group_name = azurerm_resource_group.default.name
+  location            = azurerm_resource_group.default.location
+
+  security_rule {
+    name                       = "AllowSSHHTTPSInbound"
+    priority                   = 400
+    direction                  = "Inbound"
+    access                     = "Allow"
+    protocol                   = "Tcp"
+    source_port_range          = "*"
+    destination_port_ranges    = [22, 80, 443]
```

Comment:
Debatable in non m2m world, but since it is internal system, do we need to keep 80 open? It redirects, yes, but what if redirect is misconfigured intentionally or non-intentionally on specific host?

Reply:
If we stick with caddy as a reverse proxy, we can probably drop port 80 here. nginx would have needed port 80 for HTTP-01 challenge, but caddy can also do TLS-ALPN-01.

Reply:
@alextserepov : Notice: nginx is still used as reverse proxy on some other hosts, so we would also have to change to caddy in other host configurations if we want to completely drop port 80.

Reply:
Like I said, this one is debatable in a non m2m world. As a security guy, I believe that all open-text protocols/ports should be disabled/closed on ingress. Keeping it open/served is asking for trouble, especially if there are/might be open-text-credentials flying. And the price we have to pay with nginx is just to type https:// the first time we open the link in a browser. After that it will keep on choosing HTTPs for this address in the same browser. Not a big deal, taking into account that CI/CD is internal. But I'm happy to stand corrected here. 100% open-minded.

@henrirosten Yes, you are correct. We should unify the system, thus if we switch to caddy here, we will have to switch on other hosts as well. 

Actually, I haven't looked into caddy myself, but I think it is a good idea to research this option. Thanks, @flokli!